torchdp/autograd_grad_sample.py

```diff
@@ -105,136 +115,160 @@ def _capture_activations(
     layer.activations = input[0].detach()
 
 
-def _capture_backprops(layer: nn.Module, _input, output):
-    """Append backprop to layer.backprops_list in backward pass."""
-    global _enforce_fresh_backprop
+def _capture_backprops(
+    layer: nn.Module,
+    _input: torch.Tensor,
+    output: torch.Tensor,
+    loss_type: str,
+    batch_dim: int,
+):
+    """Capture backprops in backward pass and store per-sample gradients."""
 
     if _hooks_disabled:
         return
 
-    if _enforce_fresh_backprop:
-        if hasattr(layer, "backprops_list"):
-            raise ValueError(
-                f"Seeing result of previous backprop, "
-                f"use clear_backprops(model) to clear"
-            )
-        _enforce_fresh_backprop = False
-
-    if not hasattr(layer, "backprops_list"):
-        layer.backprops_list = []
-    layer.backprops_list.append(output[0].detach())
+    backprops = output[0].detach()
+    _compute_grad_sample(layer, backprops, loss_type, batch_dim)
+    if hasattr(layer, "activations"):
+        del layer.activations
 
 
-def clear_backprops(model: nn.Module) -> None:
-    """Delete layer.backprops_list in every layer."""
+def clear_grad_sample(model: nn.Module) -> None:
+    """Delete 'param.grad_sample' in every parameter of the model."""
     for layer in model.modules():
-        if hasattr(layer, "backprops_list"):
-            del layer.backprops_list
+        for param in layer.parameters():
+            if hasattr(param, "grad_sample"):
+                del param.grad_sample
+
+
+def _create_or_extend_grad_sample(
+    param: torch.Tensor, grad_sample: torch.Tensor, batch_dim: int
+) -> None:
+    """
+    Create a 'grad_sample' attribute in the given parameter, or append to it
+    if it already exists.
+    """
+
+    if hasattr(param, "grad_sample"):
+        param.grad_sample = torch.cat((param.grad_sample, grad_sample), batch_dim)
```

Comment:
This gets a little annoying because now param.grad_sample can be either a tensor or a list. Then, when we call .clip(), before we use .grad_sample we'd have to first check whether it's a list and if so call torch.cat(). Definitely doable though.

Reply:
we could have grad_sample_list for that purpose. Anyhow, Since this is not happening too often (we might extend at most 4 times and the do accumulation) I am not insisting on it as the perf gain is probably negligible.